--- strace-4.5.15/linux/ioctlent.sh
+++ strace-4.5.15/linux/ioctlent.sh
@@ -78,12 +78,48 @@
 	>> ioctls.h
 
 # Some use a special base to offset their ioctls on. Extract that as well.
+# Some use 2 defines: _IOC(_IOC_NONE,DM_IOCTL,DM_LIST_DEVICES_CMD,....)
 : > ioctldefs.h
 
-bases=$(sed -ne 's/.*_IOC_NONE.*,[[:space:]]*\([A-Z][A-Z0-9_]\+\)[[:space:]+,].*/\1/p' ioctls.h | uniq | sort)
+bases=$(sed -n \
+	-e 's/.*_IOC_NONE.*,[[:space:]]*\([A-Z][A-Z0-9_]\+\)[[:space:]]*,[[:space:]]*\([A-Z][A-Z0-9_]\+\)[[:space:]+,].*/\1\n\2/p' \
+	-e 's/.*_IOC_NONE.*,[[:space:]]*\([A-Z][A-Z0-9_]\+\)[[:space:]+,].*/\1/p' \
+	ioctls.h | sort -u)
 for base in $bases ; do
-	echo "Looking for $base"
+	printf "Looking for $base"
 	regexp="^[[:space:]]*#[[:space:]]*define[[:space:]]\+$base"
 	(cd $dir ; grep -h $regexp 2>/dev/null $files ) | \
 		grep -v '\<_IO' >> ioctldefs.h
+
+	if ! grep -qs "\<$base\>" ioctldefs.h ; then
+		# DM_* stuff uses an enum
+		(
+		cd $dir
+		${CPP:-cpp} -E -dD -P $(grep -l $base $files 2>/dev/null) | \
+		awk -v base="$base" '{
+			if ($1 == "enum") {
+				val = 0
+				while ($NF != "};") {
+					if (!getline)
+						exit
+					gsub(/,/, "")
+					if ($0 ~ /=/)
+						val = $NF
+					if ($1 == base) {
+						print "#define " base " " val
+						exit
+					}
+					val++
+				}
+			}
+		}'
+		) >> ioctldefs.h
+		if ! grep -qs "\<$base\>" ioctldefs.h ; then
+			echo ": FAIL"
+		else
+			echo " (enum!)"
+		fi
+	else
+		echo
+	fi
 done
--- strace-4.5.15/linux/ioctlsort.c
+++ strace-4.5.15/linux/ioctlsort.c
@@ -35,9 +35,9 @@
 int main(int argc, char** argv) {
 	int i;
 
-#ifdef POWERPC			/* unspeakable kludge */
+#ifdef __powerpc__			/* unspeakable kludge */
 	for (i = 0; i < nioctls; i++)
-		ioctls[i].code &= ~_IOC_DIRMASK;
+		ioctls[i].code &= ~(1 << 29);
 #endif
 
 	qsort(ioctls, nioctls, sizeof(ioctls[0]), compare);
--- strace-4.5.15/signalent.sh
+++ strace-4.5.15/signalent.sh
@@ -28,7 +28,7 @@
 
 cat $* |
 	sed -n -e 's/\/\*.*\*\// /' -e 's/^#[ 	]*define[ 	][ 	]*SIG\([^_ 	]*\)[ 	][ 	]*\([0-9][0-9]*\)[ 	]*$/\1 \2/p' |
-	sort +1n |
+	sort -k2n |
 	awk '
 	BEGIN {
 		tabs = "\t\t\t\t\t\t\t\t"
--- strace-4.5.15/syscallent.sh
+++ strace-4.5.15/syscallent.sh
@@ -30,7 +30,7 @@
 	sed -n 's/^#[ 	]*define[ 	][ 	]*SYS_\([^ 	]*\)[ 	]*[^0-9]*\([0-9]*\).*$/\1 \2/p
 s/^#[ 	]*define[ 	][ 	]*__NR_\([^ 	]*\)[ 	]*[^0-9]*\([0-9]*\).*$/\1 \2/p
 s/^#[ ]*define[ ][ ]*__NR_\([^ ]*\)[ ]*[^0-9()]*(__NR_Linux + \([0-9]*\))$/\1 \2/p' |
-	sort +1n | uniq |
+	sort -k2n | uniq |
 	awk '
 	BEGIN {
 		tabs = "\t\t\t\t\t\t\t\t"
@@ -59,6 +59,7 @@
 		print s
 	}
 	END {
+		exit
 		limit = call + 100
 		while (++call < limit) {
 			f = "printargs"
