Index: git/linux-2.6/mm/bootmem.c
===================================================================
--- git.orig/linux-2.6/mm/bootmem.c	2005-12-19 11:38:01.000000000 +0800
+++ git/linux-2.6/mm/bootmem.c	2005-12-19 11:38:20.000000000 +0800
@@ -307,7 +307,7 @@
 				if (j + 16 < BITS_PER_LONG)
 					prefetchw(page + j + 16);
 				__ClearPageReserved(page + j);
-				set_page_count(page + j, 0);
+				/*set_page_count(page + j, 0);*/
 			}
 			__free_pages(page, order);
 			i += BITS_PER_LONG;
