Index: git/linux-2.6/Makefile
===================================================================
--- git.orig/linux-2.6/Makefile	2005-12-21 10:08:20.000000000 +0800
+++ git/linux-2.6/Makefile	2005-12-21 10:08:25.000000000 +0800
@@ -588,12 +588,12 @@
 
 vmlinux-dirs	:= $(patsubst %/,%,$(filter %/, $(init-y) $(init-m) \
 		     $(core-y) $(core-m) $(drivers-y) $(drivers-m) \
-		     $(net-y) $(net-m) $(libs-y) $(libs-m)))
+		     $(net-y) $(net-m) $(libs-y) $(libs-m) $(EXTRA_MODULE_DIRS)))
 
 vmlinux-alldirs	:= $(sort $(vmlinux-dirs) $(patsubst %/,%,$(filter %/, \
 		     $(init-n) $(init-) \
 		     $(core-n) $(core-) $(drivers-n) $(drivers-) \
-		     $(net-n)  $(net-)  $(libs-n)    $(libs-))))
+		     $(net-n)  $(net-)  $(libs-n)    $(libs-) $(EXTRA_MODULE_DIRS))))
 
 init-y		:= $(patsubst %/, %/built-in.o, $(init-y))
 core-y		:= $(patsubst %/, %/built-in.o, $(core-y))
@@ -951,11 +951,11 @@
 ifeq "$(strip $(INSTALL_MOD_PATH))" ""
 depmod_opts	:=
 else
-depmod_opts	:= -b $(INSTALL_MOD_PATH) -r
+depmod_opts	:= -b $(INSTALL_MOD_PATH)/lib/modules -r
 endif
 .PHONY: _modinst_post
 _modinst_post: _modinst_
-	if [ -r System.map -a -x $(DEPMOD) ]; then $(DEPMOD) -ae -F System.map $(depmod_opts) $(KERNELRELEASE); fi
+	if [ -r System.map -a -x $(word 1, $(DEPMOD)) ]; then $(DEPMOD) -ae -F System.map $(depmod_opts) $(KERNELRELEASE); fi
 
 else # CONFIG_MODULES
 
Index: git/linux-2.6/fs/namei.c
===================================================================
--- git.orig/linux-2.6/fs/namei.c	2005-12-21 10:08:20.000000000 +0800
+++ git/linux-2.6/fs/namei.c	2005-12-21 10:08:25.000000000 +0800
@@ -117,12 +117,14 @@
 	int retval;
 	unsigned long len = PATH_MAX;
 
+#ifdef CONFIG_MMU
 	if (!segment_eq(get_fs(), KERNEL_DS)) {
 		if ((unsigned long) filename >= TASK_SIZE)
 			return -EFAULT;
 		if (TASK_SIZE - (unsigned long) filename < PATH_MAX)
 			len = TASK_SIZE - (unsigned long) filename;
 	}
+#endif
 
 	retval = strncpy_from_user(page, filename, len);
 	if (retval > 0) {
Index: git/linux-2.6/init/Kconfig
===================================================================
--- git.orig/linux-2.6/init/Kconfig	2005-12-21 10:08:20.000000000 +0800
+++ git/linux-2.6/init/Kconfig	2005-12-21 10:08:25.000000000 +0800
@@ -403,6 +403,7 @@
 
 config TINY_SHMEM
 	default !SHMEM
+	depends MMU
 	bool
 
 config BASE_SMALL
Index: git/linux-2.6/scripts/Makefile.modinst
===================================================================
--- git.orig/linux-2.6/scripts/Makefile.modinst	2005-12-21 10:08:20.000000000 +0800
+++ git/linux-2.6/scripts/Makefile.modinst	2005-12-21 10:08:25.000000000 +0800
@@ -23,7 +23,8 @@
 INSTALL_MOD_DIR ?= extra
 ext-mod-dir = $(INSTALL_MOD_DIR)$(subst $(KBUILD_EXTMOD),,$(@D))
 
-modinst_dir = $(if $(KBUILD_EXTMOD),$(ext-mod-dir),kernel/$(@D))
+#modinst_dir = $(if $(KBUILD_EXTMOD),$(ext-mod-dir),kernel/$(@D))
+modinst_dir = $(subst $(dir $(TOPDIR))modules/,,$(if $(KBUILD_EXTMOD),$(ext-mod-dir),kernel/$(@D)))
 
 $(modules):
 	$(call cmd,modules_install,$(MODLIB)/$(modinst_dir))
Index: git/linux-2.6/drivers/mtd/maps/uclinux.c
===================================================================
--- git.orig/linux-2.6/drivers/mtd/maps/uclinux.c	2005-12-21 10:08:20.000000000 +0800
+++ git/linux-2.6/drivers/mtd/maps/uclinux.c	2005-12-21 10:08:25.000000000 +0800
@@ -34,7 +34,20 @@
 /****************************************************************************/
 
 struct mtd_partition uclinux_romfs[] = {
-	{ .name = "ROMfs" }
+	{
+#if defined CONFIG_EXT2_FS
+	.name = "EXT2fs",
+#elif defined  CONFIG_EXT3_FS
+	.name = "EXT3fs",
+#elif defined  CONFIG_ROMFS_FS
+	.name = "ROMfs" ,
+#elif defined  CONFIG_CRAMFS
+	.name = "CRAMfs",
+#elif defined CONFIG_JFFS2_FS
+	.name = "JFFS2fs",
+#endif
+
+	}
 };
 
 #define	NUM_PARTITIONS	(sizeof(uclinux_romfs) / sizeof(uclinux_romfs[0]))
@@ -52,16 +65,28 @@
 
 /****************************************************************************/
 
+extern unsigned long memory_mtd_start;
+
 int __init uclinux_mtd_init(void)
 {
 	struct mtd_info *mtd;
 	struct map_info *mapp;
+#ifdef CONFIG_BFIN
+	int i;
+	unsigned long addr = (unsigned long) memory_mtd_start;
+#else
 	extern char _ebss;
 	unsigned long addr = (unsigned long) &_ebss;
+#endif
 
 	mapp = &uclinux_ram_map;
 	mapp->phys = addr;
 	mapp->size = PAGE_ALIGN(ntohl(*((unsigned long *)(addr + 8))));
+
+#if defined(CONFIG_EXT2_FS) || defined(CONFIG_EXT3_FS)
+	mapp->size = *((unsigned long *)(addr + 0x404)) * 1024;
+#endif
+
 	mapp->bankwidth = 4;
 
 	printk("uclinux[mtd]: RAM probe address=0x%x size=0x%x\n",
@@ -86,9 +111,18 @@
 	mtd->owner = THIS_MODULE;
 	mtd->point = uclinux_point;
 	mtd->priv = mapp;
+	++mtd->usecount;
 
 	uclinux_ram_mtdinfo = mtd;
-	add_mtd_partitions(mtd, uclinux_romfs, NUM_PARTITIONS);
+#ifdef CONFIG_MTD_PARTITIONS
+	i = add_mtd_partitions(mtd, uclinux_romfs, NUM_PARTITIONS);
+#else
+	i = add_mtd_device(mtd);
+#endif
+	if (i) {
+		printk("uclinux[mtd]: failed to add mtd device\n");
+		return i;
+	}
 
 	printk("uclinux[mtd]: set %s to be root filesystem\n",
 	     	uclinux_romfs[0].name);
@@ -102,7 +136,11 @@
 void __exit uclinux_mtd_cleanup(void)
 {
 	if (uclinux_ram_mtdinfo) {
+#ifdef CONFIG_MTD_PARTITIONS
 		del_mtd_partitions(uclinux_ram_mtdinfo);
+#else
+	        del_mtd_device(uclinux_ram_mtdinfo);
+#endif
 		map_destroy(uclinux_ram_mtdinfo);
 		uclinux_ram_mtdinfo = NULL;
 	}
Index: git/linux-2.6/mm/nommu.c
===================================================================
--- git.orig/linux-2.6/mm/nommu.c	2005-12-21 10:08:20.000000000 +0800
+++ git/linux-2.6/mm/nommu.c	2005-12-21 10:08:25.000000000 +0800
@@ -57,7 +57,8 @@
 EXPORT_SYMBOL(vfree);
 EXPORT_SYMBOL(vmalloc_to_page);
 EXPORT_SYMBOL(vmalloc_32);
-
+EXPORT_SYMBOL(vmap);
+EXPORT_SYMBOL(vunmap);
 /*
  * Handle all mappings that got truncated by a "truncate()"
  * system call.
