diff -Nurp --exclude=.svn xenomai-2.4.x-clean/configure xenomai-2.4.x/configure
--- xenomai-2.4.x-clean/configure	2008-12-04 19:21:40.000000000 +0800
+++ xenomai-2.4.x/configure	2009-02-23 17:12:14.000000000 +0800
@@ -2081,11 +2081,11 @@ else
 	XENO_EMPTY_CFLAGS=false
 fi
 
-if eval test $includedir = /usr/include; then
-  { { echo "$as_me:$LINENO: error: Using /usr/include as includedir is not supported. Please change your --prefix or specify another --includedir" >&5
-echo "$as_me: error: Using /usr/include as includedir is not supported. Please change your --prefix or specify another --includedir" >&2;}
-   { (exit 1); exit 1; }; }
-fi
+#if eval test $includedir = /usr/include; then
+#  { { echo "$as_me:$LINENO: error: Using /usr/include as includedir is not supported. Please change your --prefix or specify another --includedir" >&5
+#echo "$as_me: error: Using /usr/include as includedir is not supported. Please change your --prefix or specify another --includedir" >&2;}
+#   { (exit 1); exit 1; }; }
+#fi
 
 ac_config_headers="$ac_config_headers src/include/xeno_config.h"

diff -Nurp --exclude=.svn xenomai-2.4.x-clean/Makefile.in xenomai-2.4.x/Makefile.in
--- xenomai-2.4.x-clean/Makefile.in	2008-06-02 05:32:28.000000000 +0800
+++ xenomai-2.4.x/Makefile.in	2009-02-23 17:12:14.000000000 +0800
@@ -671,7 +671,7 @@ dist-hook:
 	test -e $(srcdir)/doc || rm -rf $(distdir)/doc
 
 dev devices:
-	@$(mkinstalldirs) $(DESTDIR)/dev; \
+	$(mkinstalldirs) $(DESTDIR)/dev; \
 	if test x`id -u` = x0; then \
            sudo= ; \
         else if test \! "x`type $(SUDO)`" = x ; then \
@@ -707,7 +707,7 @@ echo '*** privileges.' ; \
           $$sudo mknod -m 666 $(DESTDIR)/dev/rtheap c 10 254 ; \
         fi
 
-install-exec-local: devices
+install-exec-local: #devices
 
 install-user:
 	$(MAKE) SUDO=false install
diff -Nurp --exclude=.svn xenomai-2.4.x-clean/src/skins/posix/Makefile.am xenomai-2.4.x/src/skins/posix/Makefile.am
--- xenomai-2.4.x-clean/src/skins/posix/Makefile.am	2008-03-10 06:33:59.000000000 +0800
+++ xenomai-2.4.x/src/skins/posix/Makefile.am	2009-02-23 17:12:14.000000000 +0800
@@ -2,7 +2,7 @@ includedir = $(prefix)/include/posix
 
 lib_LTLIBRARIES = libpthread_rt.la
 
-libpthread_rt_la_LDFLAGS = -version-info 1:0:0 -lpthread
+libpthread_rt_la_LDFLAGS = -version-info 1:0:0
 
 libpthread_rt_la_SOURCES = \
 	init.c \
diff -Nurp --exclude=.svn xenomai-2.4.x-clean/src/skins/posix/Makefile.in xenomai-2.4.x/src/skins/posix/Makefile.in
--- xenomai-2.4.x-clean/src/skins/posix/Makefile.in	2008-06-02 05:32:28.000000000 +0800
+++ xenomai-2.4.x/src/skins/posix/Makefile.in	2009-02-23 17:12:14.000000000 +0800
@@ -239,7 +239,7 @@ target_vendor = @target_vendor@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
 lib_LTLIBRARIES = libpthread_rt.la
-libpthread_rt_la_LDFLAGS = -version-info 1:0:0 -lpthread
+libpthread_rt_la_LDFLAGS = -version-info 1:0:0
 libpthread_rt_la_SOURCES = \
 	init.c \
 	thread.c \
diff -Nurp --exclude=.svn xenomai-2.4.x-clean/src/testsuite/clocktest/Makefile.am xenomai-2.4.x/src/testsuite/clocktest/Makefile.am
--- xenomai-2.4.x-clean/src/testsuite/clocktest/Makefile.am	2008-03-01 04:41:56.000000000 +0800
+++ xenomai-2.4.x/src/testsuite/clocktest/Makefile.am	2009-02-23 17:12:14.000000000 +0800
@@ -9,7 +9,7 @@ clocktest_CPPFLAGS = -I$(top_srcdir)/inc
 clocktest_LDFLAGS = $(XENO_POSIX_WRAPPERS) $(XENO_USER_LDFLAGS)
 
 clocktest_LDADD = \
-	../../skins/posix/libpthread_rt.la -lpthread -lrt
+	-lpthread -lrt ../../skins/posix/libpthread_rt.la
 
 install-data-local:
 	$(mkinstalldirs) $(DESTDIR)$(testdir)
diff -Nurp --exclude=.svn xenomai-2.4.x-clean/src/testsuite/clocktest/Makefile.in xenomai-2.4.x/src/testsuite/clocktest/Makefile.in
--- xenomai-2.4.x-clean/src/testsuite/clocktest/Makefile.in	2008-06-02 05:32:28.000000000 +0800
+++ xenomai-2.4.x/src/testsuite/clocktest/Makefile.in	2009-02-23 17:12:14.000000000 +0800
@@ -232,7 +232,7 @@ clocktest_SOURCES = clocktest.c
 clocktest_CPPFLAGS = -I$(top_srcdir)/include/posix $(XENO_USER_CFLAGS) -I$(top_srcdir)/include
 clocktest_LDFLAGS = $(XENO_POSIX_WRAPPERS) $(XENO_USER_LDFLAGS)
 clocktest_LDADD = \
-	../../skins/posix/libpthread_rt.la -lpthread -lrt
+	-lpthread -lrt ../../skins/posix/libpthread_rt.la
 
 EXTRA_DIST = runinfo.in
 all: all-am
diff -Nurp --exclude=.svn xenomai-2.4.x-clean/src/testsuite/cyclic/Makefile.am xenomai-2.4.x/src/testsuite/cyclic/Makefile.am
--- xenomai-2.4.x-clean/src/testsuite/cyclic/Makefile.am	2008-03-01 04:41:56.000000000 +0800
+++ xenomai-2.4.x/src/testsuite/cyclic/Makefile.am	2009-02-23 17:12:14.000000000 +0800
@@ -9,7 +9,7 @@ cyclictest_CPPFLAGS = -I$(top_srcdir)/in
 cyclictest_LDFLAGS = $(XENO_POSIX_WRAPPERS) $(XENO_USER_LDFLAGS)
 
 cyclictest_LDADD = \
-	../../skins/posix/libpthread_rt.la -lpthread -lrt
+	-lpthread -lrt ../../skins/posix/libpthread_rt.la
 
 install-data-local:
 	$(mkinstalldirs) $(DESTDIR)$(testdir)
diff -Nurp --exclude=.svn xenomai-2.4.x-clean/src/testsuite/cyclic/Makefile.in xenomai-2.4.x/src/testsuite/cyclic/Makefile.in
--- xenomai-2.4.x-clean/src/testsuite/cyclic/Makefile.in	2008-06-02 05:32:28.000000000 +0800
+++ xenomai-2.4.x/src/testsuite/cyclic/Makefile.in	2009-02-23 17:12:14.000000000 +0800
@@ -232,7 +232,7 @@ cyclictest_SOURCES = cyclictest.c
 cyclictest_CPPFLAGS = -I$(top_srcdir)/include/posix $(XENO_USER_CFLAGS) -DIPIPE_TRACE=1 -I$(top_srcdir)/include
 cyclictest_LDFLAGS = $(XENO_POSIX_WRAPPERS) $(XENO_USER_LDFLAGS)
 cyclictest_LDADD = \
-	../../skins/posix/libpthread_rt.la -lpthread -lrt
+	-lpthread -lrt ../../skins/posix/libpthread_rt.la
 
 EXTRA_DIST = runinfo.in
 all: all-am
diff -Nurp --exclude=.svn xenomai-2.4.x-clean/src/testsuite/irqbench/Makefile.am xenomai-2.4.x/src/testsuite/irqbench/Makefile.am
--- xenomai-2.4.x-clean/src/testsuite/irqbench/Makefile.am	2008-03-01 04:41:56.000000000 +0800
+++ xenomai-2.4.x/src/testsuite/irqbench/Makefile.am	2009-02-23 17:12:14.000000000 +0800
@@ -19,7 +19,7 @@ irqloop_LDFLAGS = \
 	$(XENO_USER_LDFLAGS)
 
 irqloop_LDADD = \
-	../../skins/posix/libpthread_rt.la -lpthread -lrt
+	-lpthread -lrt ../../skins/posix/libpthread_rt.la
 
 
 irqbench_SOURCES = irqbench.c
diff -Nurp --exclude=.svn xenomai-2.4.x-clean/src/testsuite/irqbench/Makefile.in xenomai-2.4.x/src/testsuite/irqbench/Makefile.in
--- xenomai-2.4.x-clean/src/testsuite/irqbench/Makefile.in	2008-06-02 05:32:28.000000000 +0800
+++ xenomai-2.4.x/src/testsuite/irqbench/Makefile.in	2009-02-23 17:12:14.000000000 +0800
@@ -247,7 +247,7 @@ irqloop_LDFLAGS = \
 	$(XENO_USER_LDFLAGS)
 
 irqloop_LDADD = \
-	../../skins/posix/libpthread_rt.la -lpthread -lrt
+	-lpthread -lrt ../../skins/posix/libpthread_rt.la
 
 irqbench_SOURCES = irqbench.c
 irqbench_CPPFLAGS = \
diff -Nurp --exclude=.svn xenomai-2.4.x-clean/src/testsuite/switchtest/Makefile.am xenomai-2.4.x/src/testsuite/switchtest/Makefile.am
--- xenomai-2.4.x-clean/src/testsuite/switchtest/Makefile.am	2008-03-01 04:41:56.000000000 +0800
+++ xenomai-2.4.x/src/testsuite/switchtest/Makefile.am	2009-02-23 17:12:14.000000000 +0800
@@ -9,7 +9,7 @@ switchtest_CPPFLAGS = -I$(top_srcdir)/in
 switchtest_LDFLAGS =  $(XENO_POSIX_WRAPPERS) $(XENO_USER_LDFLAGS)
 
 switchtest_LDADD = \
-	../../skins/posix/libpthread_rt.la -lpthread -lrt
+	-lpthread -lrt ../../skins/posix/libpthread_rt.la
 
 install-data-local:
 	$(mkinstalldirs) $(DESTDIR)$(testdir)
diff -Nurp --exclude=.svn xenomai-2.4.x-clean/src/testsuite/switchtest/Makefile.in xenomai-2.4.x/src/testsuite/switchtest/Makefile.in
--- xenomai-2.4.x-clean/src/testsuite/switchtest/Makefile.in	2008-06-02 05:32:28.000000000 +0800
+++ xenomai-2.4.x/src/testsuite/switchtest/Makefile.in	2009-02-23 17:12:14.000000000 +0800
@@ -232,7 +232,7 @@ switchtest_SOURCES = switchtest.c
 switchtest_CPPFLAGS = -I$(top_srcdir)/include/posix $(XENO_USER_CFLAGS) -g -I$(top_srcdir)/include
 switchtest_LDFLAGS = $(XENO_POSIX_WRAPPERS) $(XENO_USER_LDFLAGS)
 switchtest_LDADD = \
-	../../skins/posix/libpthread_rt.la -lpthread -lrt
+	-lpthread -lrt ../../skins/posix/libpthread_rt.la
 
 EXTRA_DIST = runinfo.in
 all: all-am
