--- speex/configure.ac
+++ speex/configure.ac
@@ -128,7 +128,9 @@
 AC_ARG_ENABLE(blackfin-asm, [  --enable-blackfin-asm      Make use of Blackfin assembly optimizations],
 [if test "$enableval" = yes; then
   AC_DEFINE([BFIN_ASM], , [Make use of Blackfin assembly optimizations])
-  LDFLAGS="-Wl,-elf2flt=-s100000"
+  case $host_os in
+    uclinux) LDFLAGS="$LDFLAGS -Wl,-elf2flt=-s100000";;
+  esac
 fi])
 
 AC_ARG_ENABLE(fixed-point-debug, [  --enable-fixed-point-debug  Debug fixed-point implementation],
