--- a/src/3rdparty/webkit/JavaScriptCore/kjs/collector.cpp
+++ b/src/3rdparty/webkit/JavaScriptCore/kjs/collector.cpp
@@ -59,7 +59,7 @@
 #ifdef __UCLIBC__
 // versions of uClibc 0.9.28 and below do not have
 // pthread_getattr_np or pthread_attr_getstack.
-#if __UCLIBC_MAJOR__ == 0 && \
+#if 1 || __UCLIBC_MAJOR__ == 0 && \
     (__UCLIBC_MINOR__ < 9 || \
      (__UCLIBC_MINOR__ == 9 && __UCLIBC_SUBLEVEL__ < 29))
 #define UCLIBC_USE_PROC_SELF_MAPS 1
