Index: testcases/kernel/syscalls/gethostid/gethostid01.c
===================================================================
RCS file: /ltp/ltp-full-20060411/testcases/kernel/syscalls/gethostid/gethostid01.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 gethostid01.c
--- testcases/kernel/syscalls/gethostid/gethostid01.c	12 Apr 2006 09:16:45 -0000	1.1.1.1
+++ testcases/kernel/syscalls/gethostid/gethostid01.c	13 Apr 2006 08:07:29 -0000
@@ -182,7 +182,11 @@
 		     TEST_ERRNO, strerror(TEST_ERRNO));
 	    continue;	/* next loop for MTKERNEL */
 	}
+#ifdef UCLINUX
+	sprintf(hostid, "%lx", TEST_RETURN);
+#else
 	sprintf(hostid, "%0.8lx", TEST_RETURN);
+#endif
 
 	/***************************************************************
 	 * only perform functional verification if flag set (-f not given)
