--- testcases/kernel/syscalls/getsockname/getsockname01.c	2005-12-15 10:59:11.000000000 +0800
+++ ltp-full-20051205.org/testcases/kernel/syscalls/getsockname/getsockname01.c	2005-12-13 11:46:38.000000000 +0800
@@ -82,15 +82,12 @@
 	{ PF_INET, SOCK_STREAM, 0, (struct sockaddr *)&fsin1,
 		&sinlen, -1, ENOTSOCK, setup0, cleanup0,
 		"bad file descriptor" },
-#ifndef UCLINUX
-	/* Ignore this test set on uClinux */
 	{ PF_INET, SOCK_STREAM, 0, (struct sockaddr *)0,
 		&sinlen, -1, EFAULT, setup1, cleanup1,
 		"invalid socket buffer" },
 	{ PF_INET, SOCK_STREAM, 0, (struct sockaddr *)&fsin1,
 		(int *)1, -1, EFAULT, setup1, cleanup1,
 		"invalid salen" },
-#endif
 };
 
 int TST_TOTAL=sizeof(tdat)/sizeof(tdat[0]); /* Total number of test cases. */
