--- ltp-full-20060105.org/testcases/kernel/syscalls/getsockopt/getsockopt01.c	2006-01-23 11:30:06.000000000 +0800
+++ testcases/kernel/syscalls/getsockopt/getsockopt01.c	2006-01-25 12:16:46.000000000 +0800
@@ -87,12 +87,14 @@
 	{ PF_INET, SOCK_STREAM, 0, SOL_SOCKET, SO_OOBINLINE, &optval, &optlen,
 		(struct sockaddr *)&fsin1, sizeof(fsin1), -1,
 		ENOTSOCK, setup0, cleanup0, "bad file descriptor" },
+#ifndef UCLINUX
 	{ PF_INET, SOCK_STREAM, 0, SOL_SOCKET, SO_OOBINLINE, 0, &optlen,
 		(struct sockaddr *)&fsin1, sizeof(fsin1), -1,
 		EFAULT, setup1, cleanup1, "invalid option buffer" },
 	{ PF_INET, SOCK_STREAM, 0, SOL_SOCKET, SO_OOBINLINE, &optval, 0,
 		(struct sockaddr *)&fsin1, sizeof(fsin1), -1,
 		EFAULT, setup1, cleanup1, "invalid optlen" },
+#endif /* UCLINUX */
 	{ PF_INET, SOCK_STREAM, 0, 500, SO_OOBINLINE, &optval, &optlen,
 		(struct sockaddr *)&fsin1, sizeof(fsin1), -1,
 		EOPNOTSUPP, setup1, cleanup1, "invalid level" },
