--- ltp-full-20060105.org/testcases/kernel/syscalls/kill/kill08.c	2006-01-23 11:30:02.000000000 +0800
+++ testcases/kernel/syscalls/kill/kill08.c	2006-01-25 12:21:26.000000000 +0800
@@ -109,7 +109,7 @@
 		} else if (pid1 == 0) {
 			setpgrp();
 			for (i = 0; i < 5; i++) {
-				pid2 = fork();
+				pid2 = FORK_OR_VFORK();
 				if (pid2 < 0) {
 					tst_brkm(TBROK, cleanup, "Fork failed");
 				} else if (pid2 == 0) {
