Index: testcases/kernel/syscalls/lstat/lstat02.c
===================================================================
RCS file: /ltp/ltp-full-20060411/testcases/kernel/syscalls/lstat/lstat02.c,v
diff -u -r1.1 -r1.2
--- testcases/kernel/syscalls/lstat/lstat02.c	12 Apr 2006 09:16:07 -0000	1.1
+++ testcases/kernel/syscalls/lstat/lstat02.c	13 Apr 2006 08:25:49 -0000	1.2
@@ -253,7 +253,11 @@
 	if (bad_addr <= 0) {
 		tst_brkm(TBROK, cleanup, "mmap failed");
 	}
+#ifdef UCLINUX
+	Test_cases[1].pathname = bad_addr;
+#else
 	Test_cases[2].pathname = bad_addr;
+#endif
 
 	/* call individual setup functions */
 	for (ind = 0; Test_cases[ind].desc != NULL; ind++) {
