Index: testcases/kernel/syscalls/mknod/mknod06.c
===================================================================
RCS file: /ltp/ltp-full-20060411/testcases/kernel/syscalls/mknod/mknod06.c,v
retrieving revision 1.1.1.1
diff -u -r1.1.1.1 mknod06.c
--- testcases/kernel/syscalls/mknod/mknod06.c	12 Apr 2006 09:16:11 -0000	1.1.1.1
+++ testcases/kernel/syscalls/mknod/mknod06.c	13 Apr 2006 09:12:42 -0000
@@ -241,7 +241,11 @@
 	if (bad_addr <= 0) {
 		tst_brkm(TBROK, cleanup, "mmap failed");
 	}
+#ifdef UCLINUX
+	Test_cases[1].pathname = bad_addr;
+#else
 	Test_cases[2].pathname = bad_addr;
+#endif
 
 	/* call individual setup functions */
 	for (ind = 0; Test_cases[ind].desc != NULL; ind++) {
