--- testcases/kernel/syscalls/readv/readv02.c	2005-12-15 15:12:14.000000000 +0800
+++ ltp-full-20051205.org/testcases/kernel/syscalls/readv/readv02.c	2005-12-13 11:46:38.000000000 +0800
@@ -132,12 +132,6 @@
 				 "value");
 		}
 
-#ifndef UCLINUX
-		/*
-		 * This test set is trying to access outside address space.
-		 * It doesn't work on uClinux.
-		 * Ignore this test set on uClinux.
-		 */
 //test2:
 		l_seek(fd[0], CHUNK * 6, 0);
 		if (readv(fd[0], (rd_iovec + 6), 3) < 0) {
@@ -156,7 +150,6 @@
 			tst_resm(TFAIL, "Error: readv returned a positive "
 				 "value");
 		}
-#endif
 
 //test3:
 		if (readv(fd[1], (rd_iovec + 9), 1) < 0) {
