Index: alsa-lib-1.0.12/src/conf/cards/AD1980.conf
===================================================================
--- alsa-lib-1.0.12/src/conf/cards/AD1980.conf	(revision 6123)
+++ alsa-lib-1.0.12/src/conf/cards/AD1980.conf	(working copy)
@@ -2,17 +2,13 @@
 	type plug
 	slave {
 		pcm {
-			type mmap_emul
-			slave {
-				pcm {
-					type hw
-					card 0
-					format S16_LE
-					rate 48000
-					channels 2
-					sync_ptr_ioctl 1
-				}
-			}
+			type hw
+			card 0
+			format S16_LE
+			rate 48000
+			channels 2
+			sync_ptr_ioctl 1
+		
 		}
 	}
 }
