Index: openobex-1.3.0/aclocal.m4
===================================================================
--- openobex-1.3.0/aclocal.m4	(revision 6563)
+++ openobex-1.3.0/aclocal.m4	(working copy)
@@ -19,11 +19,15 @@
 AC_DEFUN([IRDA_HOOK],[
 	AC_CACHE_CHECK([for IrDA support],am_cv_irda_found,[
 
+		save_CPPFLAGS="$CPPFLAGS"
+		CPPFLAGS="$CPPFLAGS -I$srcdir"
 		AC_TRY_COMPILE([#include <sys/socket.h>
-				#include "../openobex-1.3.0/src/irda.h"],
+				#include "src/irda.h"],
 		[struct irda_device_list l;],
 		am_cv_irda_found=yes,
-		am_cv_irda_found=no)])
+		am_cv_irda_found=no)
+		CPPFLAGS="$save_CPPFLAGS"
+		])
 
 		if test $am_cv_irda_found = yes; then
 			AC_DEFINE(HAVE_IRDA,1,[Define if system supports IrDA])
@@ -46,7 +50,6 @@
 
 AC_DEFUN([BLUETOOTH_HOOK],[
 	AC_CACHE_CHECK([for Bluetooth support],am_cv_bluetooth_found,[
-
 		AC_TRY_COMPILE([#include <sys/socket.h>
 				#include <bluetooth/bluetooth.h>
 				#include <bluetooth/rfcomm.h>],
