should be converted to a "no-nls" make flag

--- procps-3.2.7/proc/escape.c
+++ procps-3.2.7/proc/escape.c
@@ -15,7 +15,7 @@
 #include "escape.h"
 #include "readproc.h"
 
-#if (__GNU_LIBRARY__ >= 6)
+#if (0)
 # include <wchar.h>
 # include <wctype.h>
 # include <stdlib.h>  /* MB_CUR_MAX */
@@ -23,7 +23,7 @@
 # include <langinfo.h>
 #endif
 
-#if (__GNU_LIBRARY__ >= 6)
+#if (0)
 static int escape_str_utf8(char *restrict dst, const char *restrict src, int bufsize, int *maxcells){
   int my_cells = 0;
   int my_bytes = 0;
@@ -123,7 +123,7 @@
   "********************************"
   "********************************";
   
-#if (__GNU_LIBRARY__ >= 6)
+#if (0)
   static int utf_init=0;
   
   if(utf_init==0){
