diff -Nur MPlayer-1.0rc1-orig/configure MPlayer-1.0rc1/configure
--- MPlayer-1.0rc1-orig/configure	2006-10-23 06:32:31.000000000 +0800
+++ MPlayer-1.0rc1/configure	2007-01-17 12:54:10.000000000 +0800
@@ -59,7 +59,7 @@
   echo >> "$TMPLOG"
   cat "$1" >> "$TMPLOG"
   echo >> "$TMPLOG"
-  echo "$_cc $CFLAGS $_inc_extra $_ld_static $_ld_extra -o $TMPO $@" >> "$TMPLOG"
+  echo "$_cc $CFLAGS $_inc_extra $_ld_static $_ld_extra -o $TMPO $@" >> "$TMPLOG"	
   rm -f "$TMPO"
   $_cc $CFLAGS $_inc_extra $_ld_static $_ld_extra -o "$TMPO" "$@" >> "$TMPLOG" 2>&1
   TMP="$?"
@@ -1437,6 +1437,16 @@
     _optimizing=''
     ;;
 
+    bfin)
+	_def_arch='#define ARCH_BFIN 1'
+	_target_arch='TARGET_ARCH_BFIN = yes'
+	iproc='bfin'
+	proc=''
+	_march=''
+	_mcpu=''
+	_optimizing=''
+    ;;
+
   *)
     echo "The architecture of your CPU ($host_arch) is not supported by this configure script"
     echo "It seems nobody has ported MPlayer to your OS or CPU type yet."
@@ -1953,6 +1963,11 @@
   --disable-amr_nb-fixed)	_amr_nb_fixed=no ;;
   --enable-amr_wb)	_amr_wb=yes	;;
   --disable-amr_wb)	_amr_wb=no	;;
+  --enable-decoder-only=*)  _libavdecoders="`echo $ac_option | cut -d '=' -f 2`" ;;
+  --enable-encoder-only=*)  _libavencoders="`echo $ac_option | cut -d '=' -f 2`" ;;
+  --enable-parser-only=*)   _libavparsers="`echo $ac_option | cut -d '=' -f 2`" ;;
+  --enable-demuxer-only=*)  _libavdemuxers="`echo $ac_option | cut -d '=' -f 2`" ;;
+  --enable-muxer-only=*)    _libavmuxers="`echo $ac_option | cut -d '=' -f 2`" ;;
   --enable-decoder=*)	_libavdecoders="$_libavdecoders `echo $ac_option | cut -d '=' -f 2`" ;;
   --disable-decoder=*)	_libavdecoders=`echo $_libavdecoders | sed "s/\`echo $ac_option | cut -d '=' -f 2\`//g"` ;;
   --enable-encoder=*)	_libavencoders="$_libavencoders `echo $ac_option | cut -d '=' -f 2`" ;;
@@ -3803,7 +3818,7 @@
     fi
   done
   #FIXME: This is ugly as it can duplicate a -I parameter..
-  _inc_extra="$_inc_extra $_inc_x11"
+  #_inc_extra="$_inc_extra $_inc_x11"	# mhfan
   if test -z "$_inc_x11" ; then
     _x11=no
     _x11_headers="no"
@@ -3829,6 +3844,7 @@
   done
 fi
 if test "$_x11" = yes ; then
+  _inc_extra="$_inc_extra $_inc_x11"	# mhfan
   _def_x11='#define HAVE_X11 1'
   _vosrc="$_vosrc x11_common.c vo_x11.c vo_xover.c"
   _vomodules="x11 xover $_vomodules"
diff -Nur MPlayer-1.0rc1-orig/libaf/af_export.c MPlayer-1.0rc1/libaf/af_export.c
--- MPlayer-1.0rc1-orig/libaf/af_export.c	2006-10-23 06:32:31.000000000 +0800
+++ MPlayer-1.0rc1/libaf/af_export.c	2007-01-17 12:54:10.000000000 +0800
@@ -112,7 +112,7 @@
     // Initialize header
     *((int*)s->mmap_area) = af->data->nch;
     *((int*)s->mmap_area + 1) = s->sz * af->data->bps * af->data->nch;
-    msync(s->mmap_area, mapsize, MS_ASYNC);
+    //msync(s->mmap_area, mapsize, MS_ASYNC);
 
     // Use test_output to return FALSE if necessary
     return af_test_output(af, (af_data_t*)arg);
diff -Nur MPlayer-1.0rc1-orig/libmpdemux/demux_mov.c MPlayer-1.0rc1/libmpdemux/demux_mov.c
--- MPlayer-1.0rc1-orig/libmpdemux/demux_mov.c	2006-10-23 06:32:31.000000000 +0800
+++ MPlayer-1.0rc1/libmpdemux/demux_mov.c	2007-01-17 13:10:26.000000000 +0800
@@ -396,7 +396,8 @@
 	  skipped += 8;
 	  // List all compatible brands
 	  for(i = 0; i < ((len-16)/4); i++) {
-	    tmp = be2me_32(stream_read_dword(demuxer->stream));
+	    tmp = stream_read_dword(demuxer->stream);
+	    tmp = be2me_32(tmp);
 	    mp_msg(MSGT_DEMUX,MSGL_V,"ISO: File Type Compatible Brand #%d: %.4s\n",i,(char *)&tmp);
 	    skipped += 4;
 	  }
diff -Nur MPlayer-1.0rc1-orig/libvo/vo_fbdev.c MPlayer-1.0rc1/libvo/vo_fbdev.c
--- MPlayer-1.0rc1-orig/libvo/vo_fbdev.c	2006-10-23 06:32:26.000000000 +0800
+++ MPlayer-1.0rc1/libvo/vo_fbdev.c	2007-01-17 12:54:10.000000000 +0800
@@ -1004,7 +1004,8 @@
 	{
 	    int x_offset=0,y_offset=0;
 	    if ((frame_buffer = (uint8_t *) mmap(0, fb_size, PROT_READ | PROT_WRITE,
-				    MAP_SHARED, fb_dev_fd, 0)) == (uint8_t *) -1) {
+				    MAP_PRIVATE, fb_dev_fd, 0)) == (uint8_t *) -1) {
+				    //MAP_SHARED, fb_dev_fd, 0)) == (uint8_t *) -1) {
 		mp_msg(MSGT_VO, MSGL_ERR, "Can't mmap %s: %s\n", fb_dev_name, strerror(errno));
 		return 1;
 	    }
diff -Nur MPlayer-1.0rc1-orig/libvo/vo_yuv4mpeg.c MPlayer-1.0rc1/libvo/vo_yuv4mpeg.c
--- MPlayer-1.0rc1-orig/libvo/vo_yuv4mpeg.c	2006-10-23 06:32:26.000000000 +0800
+++ MPlayer-1.0rc1/libvo/vo_yuv4mpeg.c	2007-01-17 12:54:10.000000000 +0800
@@ -40,6 +40,7 @@
 
 #include "fastmemcpy.h"
 #include "libswscale/swscale.h"
+#include "libswscale/rgb2rgb.h"
 #include "libmpcodecs/vf_scale.h"
 
 static vo_info_t info = 
