diff -Nur live/config.bfin_uclinux live-new/config.bfin_uclinux
--- live/config.bfin_uclinux	2007-01-18 04:44:26.000000000 +0800
+++ live-new/config.bfin_uclinux	1970-01-01 08:00:00.000000000 +0800
@@ -1,18 +0,0 @@
-CROSS_COMPILER=        bfin-uclinux-
-COMPILE_OPTS =        $(INCLUDES) -I. -DSOCKLEN_T=socklen_t -D_LARGEFILE_SOURCE=1 -DUCLINUX -D_FILE_OFFSET_BITS=64
-C =            c
-C_COMPILER =        $(CROSS_COMPILER)gcc
-C_FLAGS =        $(COMPILE_OPTS) -Wall
-CPP =            cpp
-CPLUSPLUS_COMPILER =    $(CROSS_COMPILER)g++
-CPLUSPLUS_FLAGS =    $(COMPILE_OPTS) -Wall
-OBJ =            o
-LINK =            $(CROSS_COMPILER)g++ -Wl,-elf2flt -o
-LINK_OPTS =        -L.
-CONSOLE_LINK_OPTS =    $(LINK_OPTS)
-LIBRARY_LINK =        $(CROSS_COMPILER)ld -o
-LIBRARY_LINK_OPTS =    $(LINK_OPTS) -r -Bstatic
-LIB_SUFFIX =            a
-LIBS_FOR_CONSOLE_APPLICATION =
-LIBS_FOR_GUI_APPLICATION =
-EXE =
diff -Nur live/config.bfin_uclinux_fdpic live-new/config.bfin_uclinux_fdpic
--- live/config.bfin_uclinux_fdpic	1970-01-01 08:00:00.000000000 +0800
+++ live-new/config.bfin_uclinux_fdpic	2007-02-08 10:54:56.000000000 +0800
@@ -0,0 +1,18 @@
+CROSS_COMPILER=        bfin-uclinux-
+COMPILE_OPTS =        $(INCLUDES) -I. -DSOCKLEN_T=socklen_t -D_LARGEFILE_SOURCE=1 -DUCLINUX -D_FILE_OFFSET_BITS=64
+C =            c
+C_COMPILER =        $(CROSS_COMPILER)gcc
+C_FLAGS =        $(COMPILE_OPTS) -Wall -mfdpic
+CPP =            cpp
+CPLUSPLUS_COMPILER =    $(CROSS_COMPILER)g++
+CPLUSPLUS_FLAGS =    $(COMPILE_OPTS) -Wall -mfdpic
+OBJ =            o
+LINK =            $(CROSS_COMPILER)g++ -o
+LINK_OPTS =        -mfdpic -L.  
+CONSOLE_LINK_OPTS =    $(LINK_OPTS)
+LIBRARY_LINK =        $(CROSS_COMPILER)ar rcu 
+LIBRARY_LINK_OPTS =   
+LIB_SUFFIX =            a
+LIBS_FOR_CONSOLE_APPLICATION =
+LIBS_FOR_GUI_APPLICATION =
+EXE =
diff -Nur live/config.bfin_uclinux_flat live-new/config.bfin_uclinux_flat
--- live/config.bfin_uclinux_flat	1970-01-01 08:00:00.000000000 +0800
+++ live-new/config.bfin_uclinux_flat	2007-02-08 17:55:36.000000000 +0800
@@ -0,0 +1,18 @@
+CROSS_COMPILER=        bfin-uclinux-
+COMPILE_OPTS =        $(INCLUDES) -I. -DSOCKLEN_T=socklen_t -D_LARGEFILE_SOURCE=1 -DUCLINUX -D_FILE_OFFSET_BITS=64
+C =            c
+C_COMPILER =        $(CROSS_COMPILER)gcc
+C_FLAGS =        $(COMPILE_OPTS) -Wall
+CPP =            cpp
+CPLUSPLUS_COMPILER =    $(CROSS_COMPILER)g++
+CPLUSPLUS_FLAGS =    $(COMPILE_OPTS) -Wall
+OBJ =            o
+LINK =            $(CROSS_COMPILER)g++ -Wl,-elf2flt -o
+LINK_OPTS =        -L.
+CONSOLE_LINK_OPTS =    $(LINK_OPTS)
+LIBRARY_LINK =        $(CROSS_COMPILER)ld -o
+LIBRARY_LINK_OPTS =    $(LINK_OPTS) -r -Bstatic
+LIB_SUFFIX =            a
+LIBS_FOR_CONSOLE_APPLICATION =
+LIBS_FOR_GUI_APPLICATION =
+EXE =
diff -Nur live/mediaServer/Makefile.tail live-new/mediaServer/Makefile.tail
--- live/mediaServer/Makefile.tail	2007-01-18 04:44:25.000000000 +0800
+++ live-new/mediaServer/Makefile.tail	2007-02-12 15:10:42.000000000 +0800
@@ -25,7 +25,9 @@
 GROUPSOCK_DIR = ../groupsock
 GROUPSOCK_LIB = $(GROUPSOCK_DIR)/libgroupsock.$(LIB_SUFFIX)
 LOCAL_LIBS =	$(LIVEMEDIA_LIB) $(GROUPSOCK_LIB) \
-		$(USAGE_ENVIRONMENT_LIB) $(BASIC_USAGE_ENVIRONMENT_LIB)
+		$(BASIC_USAGE_ENVIRONMENT_LIB) \
+		$(USAGE_ENVIRONMENT_LIB) 
+		
 LIBS =			$(LOCAL_LIBS) $(LIBS_FOR_CONSOLE_APPLICATION)
 
 live555MediaServer$(EXE):	$(MEDIA_SERVER_OBJS) $(LOCAL_LIBS)
diff -Nur live/testProgs/Makefile.tail live-new/testProgs/Makefile.tail
--- live/testProgs/Makefile.tail	2007-01-18 04:44:25.000000000 +0800
+++ live-new/testProgs/Makefile.tail	2007-02-12 15:10:57.000000000 +0800
@@ -59,7 +59,8 @@
 GROUPSOCK_DIR = ../groupsock
 GROUPSOCK_LIB = $(GROUPSOCK_DIR)/libgroupsock.$(LIB_SUFFIX)
 LOCAL_LIBS =	$(LIVEMEDIA_LIB) $(GROUPSOCK_LIB) \
-		$(USAGE_ENVIRONMENT_LIB) $(BASIC_USAGE_ENVIRONMENT_LIB)
+		$(BASIC_USAGE_ENVIRONMENT_LIB) \
+		$(USAGE_ENVIRONMENT_LIB) 
 LIBS =			$(LOCAL_LIBS) $(LIBS_FOR_CONSOLE_APPLICATION)
 
 testMP3Streamer$(EXE):	$(MP3_STREAMER_OBJS) $(LOCAL_LIBS)
