diff -Nur -x 'Makefile*' -x.libs -x'.*' -x'*.bak' -x '*.prl' -x '*.status' -x '*.pri' qtopia-core-opensource-src-4.2.2/mkspecs/qws/linux-bfin-g++/qmake.conf qtopia-core-opensource-src-4.2.2-newest-toolchain/mkspecs/qws/linux-bfin-g++/qmake.conf
--- qtopia-core-opensource-src-4.2.2/mkspecs/qws/linux-bfin-g++/qmake.conf	1970-01-01 01:00:00.000000000 +0100
+++ qtopia-core-opensource-src-4.2.2-newest-toolchain/mkspecs/qws/linux-bfin-g++/qmake.conf	2007-06-19 15:22:48.000000000 +0200
@@ -0,0 +1,91 @@
+#
+# qmake configuration for linux-g++
+#
+
+MAKEFILE_GENERATOR	= UNIX
+TEMPLATE		= app
+CONFIG			+= qt link_prl
+QT                      += core gui
+QMAKE_INCREMENTAL_STYLE = sublib
+
+QMAKE_CC		= bfin-linux-uclibc-gcc
+QMAKE_LEX		= flex
+QMAKE_LEXFLAGS		=
+QMAKE_YACC		= yacc
+QMAKE_YACCFLAGS		= -d
+QMAKE_CFLAGS		= -mfdpic -pipe -D__linux__ -DQT_NOMMU -DQ_NO_SEMAPHORE -DQT_QWS_NO_SHM -DQT_NO_QWS_MULTIPROCESS
+QMAKE_CFLAGS_WARN_ON	= -Wall -W
+QMAKE_CFLAGS_WARN_OFF	=
+QMAKE_CFLAGS_RELEASE	= -O2
+QMAKE_CFLAGS_DEBUG	= -g
+QMAKE_CFLAGS_SHLIB	= 
+QMAKE_CFLAGS_YACC	= -Wno-unused -Wno-parentheses
+QMAKE_CFLAGS_THREAD	= -D_REENTRANT
+QMAKE_CFLAGS_HIDESYMS   = -fvisibility=hidden
+
+QMAKE_CXX		= bfin-linux-uclibc-g++
+QMAKE_CXXFLAGS		= $$QMAKE_CFLAGS -DQWS -fno-exceptions -fno-rtti
+QMAKE_CXXFLAGS_WARN_ON	= $$QMAKE_CFLAGS_WARN_ON
+QMAKE_CXXFLAGS_WARN_OFF	= $$QMAKE_CFLAGS_WARN_OFF
+QMAKE_CXXFLAGS_RELEASE	= $$QMAKE_CFLAGS_RELEASE
+QMAKE_CXXFLAGS_DEBUG	= $$QMAKE_CFLAGS_DEBUG
+QMAKE_CXXFLAGS_SHLIB	= $$QMAKE_CFLAGS_SHLIB
+QMAKE_CXXFLAGS_YACC	= $$QMAKE_CFLAGS_YACC
+QMAKE_CXXFLAGS_THREAD	= $$QMAKE_CFLAGS_THREAD
+QMAKE_CXXFLAGS_HIDESYMS = $$QMAKE_CFLAGS_HIDESYMS -fvisibility-inlines-hidden
+
+QMAKE_INCDIR		=
+QMAKE_LIBDIR		=
+QMAKE_INCDIR_X11	= 
+QMAKE_LIBDIR_X11	= 
+QMAKE_INCDIR_QT		= $$[QT_INSTALL_HEADERS]
+QMAKE_LIBDIR_QT		= $$[QT_INSTALL_LIBS]
+QMAKE_INCDIR_OPENGL	= /usr/X11R6/include
+QMAKE_LIBDIR_OPENGL	= /usr/X11R6/lib
+QMAKE_INCDIR_QTOPIA	= $(QPEDIR)/include/
+QMAKE_LIBDIR_QTOPIA	= $(QPEDIR)/lib
+
+QMAKE_LINK		= bfin-linux-uclibc-g++ 
+QMAKE_LINK_SHLIB	= bfin-linux-uclibc-g++
+QMAKE_LFLAGS		= 
+QMAKE_LFLAGS_RELEASE	= # -mid-shared-library -mshared-library-id=0
+QMAKE_LFLAGS_DEBUG	=
+QMAKE_LFLAGS_SHLIB      = -shared 
+QMAKE_LFLAGS_PLUGIN     = $$QMAKE_LFLAGS_SHLIB
+QMAKE_LFLAGS_SONAME     = -Wl,-soname,
+QMAKE_LFLAGS_THREAD     =
+QMAKE_RPATH             = -Wl,-rpath,
+
+QMAKE_LIBS		= 
+QMAKE_LIBS_DYNLOAD      = -ldl
+QMAKE_LIBS_X11		= 
+QMAKE_LIBS_X11SM	= 
+QMAKE_LIBS_QT		= -lqte
+QMAKE_LIBS_QT_THREAD    = -lqte-mt
+QMAKE_LIBS_QT_OPENGL	= -lqgl
+QMAKE_LIBS_QTOPIA	= -lqtopia -lqpe
+QMAKE_LIBS_THREAD       = -lpthread
+
+QMAKE_MOC		= $$[QT_INSTALL_BINS]/moc
+QMAKE_UIC		= $$[QT_INSTALL_BINS]/uic
+
+QMAKE_AR		= bfin-linux-uclibc-ar cqs
+QMAKE_RANLIB		=
+
+QMAKE_TAR		= tar -cf
+QMAKE_GZIP		= gzip -9f
+
+QMAKE_COPY		= cp -f
+QMAKE_MOVE		= mv -f
+QMAKE_DEL_FILE		= rm -f
+QMAKE_DEL_DIR		= rmdir
+QMAKE_STRIP             = strip
+QMAKE_CHK_DIR_EXISTS	= test -d
+QMAKE_MKDIR		= mkdir -p
+include(../../common/unix.conf)
+load(qt_config)
+#/home/f/blackfin-toolchain-06r2/opt/linux-uclibc/bfin-linux-uclibc/bfin-linux-uclibc/lib/libdl_pic.a
+#/home/f/blackfin-toolchain-06r2/opt/linux-uclibc/bfin-linux-uclibc/bfin-linux-uclibc/lib/mfdpic/libdl.a
+#/home/f/blackfin-toolchain-06r2/opt/linux-uclibc/bfin-linux-uclibc/bfin-linux-uclibc/lib/mfdpic/libdl.so
+#/home/f/blackfin-toolchain-06r2/opt/linux-uclibc/bfin-linux-uclibc/bfin-linux-uclibc/runtime/lib/libdl-0.9.27.so
+#/home/f/blackfin-toolchain-06r2/opt/linux-uclibc/bfin-linux-uclibc/bfin-linux-uclibc/runtime/lib/libdl.so.0
diff -Nur -x 'Makefile*' -x.libs -x'.*' -x'*.bak' -x '*.prl' -x '*.status' -x '*.pri' qtopia-core-opensource-src-4.2.2/mkspecs/qws/linux-bfin-g++/qmake.shared.conf qtopia-core-opensource-src-4.2.2-newest-toolchain/mkspecs/qws/linux-bfin-g++/qmake.shared.conf
--- qtopia-core-opensource-src-4.2.2/mkspecs/qws/linux-bfin-g++/qmake.shared.conf	1970-01-01 01:00:00.000000000 +0100
+++ qtopia-core-opensource-src-4.2.2-newest-toolchain/mkspecs/qws/linux-bfin-g++/qmake.shared.conf	2007-01-03 16:07:32.000000000 +0100
@@ -0,0 +1,91 @@
+#
+# qmake configuration for linux-g++
+#
+
+MAKEFILE_GENERATOR	= UNIX
+TEMPLATE		= app
+CONFIG			+= qt link_prl
+QT                      += core gui network
+QMAKE_INCREMENTAL_STYLE = sublib
+
+QMAKE_CC		= bfin-uclinux-gcc
+QMAKE_LEX		= flex
+QMAKE_LEXFLAGS		=
+QMAKE_YACC		= yacc
+QMAKE_YACCFLAGS		= -d
+QMAKE_CFLAGS		= -mfdpic -pipe -D__linux__ -DQT_NOMMU -DQ_NO_SEMAPHORE -DQT_QWS_NO_SHM -DQT_NO_QWS_MULTIPROCESS
+QMAKE_CFLAGS_WARN_ON	= -Wall -W
+QMAKE_CFLAGS_WARN_OFF	=
+QMAKE_CFLAGS_RELEASE	= -O2
+QMAKE_CFLAGS_DEBUG	= -g
+QMAKE_CFLAGS_SHLIB	= 
+QMAKE_CFLAGS_YACC	= -Wno-unused -Wno-parentheses
+QMAKE_CFLAGS_THREAD	= -D_REENTRANT
+QMAKE_CFLAGS_HIDESYMS   = -fvisibility=hidden
+
+QMAKE_CXX		= bfin-uclinux-g++
+QMAKE_CXXFLAGS		= $$QMAKE_CFLAGS -DQWS -fno-exceptions -fno-rtti
+QMAKE_CXXFLAGS_WARN_ON	= $$QMAKE_CFLAGS_WARN_ON
+QMAKE_CXXFLAGS_WARN_OFF	= $$QMAKE_CFLAGS_WARN_OFF
+QMAKE_CXXFLAGS_RELEASE	= $$QMAKE_CFLAGS_RELEASE
+QMAKE_CXXFLAGS_DEBUG	= $$QMAKE_CFLAGS_DEBUG
+QMAKE_CXXFLAGS_SHLIB	= $$QMAKE_CFLAGS_SHLIB
+QMAKE_CXXFLAGS_YACC	= $$QMAKE_CFLAGS_YACC
+QMAKE_CXXFLAGS_THREAD	= $$QMAKE_CFLAGS_THREAD
+QMAKE_CXXFLAGS_HIDESYMS = $$QMAKE_CFLAGS_HIDESYMS -fvisibility-inlines-hidden
+
+QMAKE_INCDIR		=
+QMAKE_LIBDIR		=
+QMAKE_INCDIR_X11	= 
+QMAKE_LIBDIR_X11	= 
+QMAKE_INCDIR_QT		= $$[QT_INSTALL_HEADERS]
+QMAKE_LIBDIR_QT		= $$[QT_INSTALL_LIBS]
+QMAKE_INCDIR_OPENGL	= /usr/X11R6/include
+QMAKE_LIBDIR_OPENGL	= /usr/X11R6/lib
+QMAKE_INCDIR_QTOPIA	= $(QPEDIR)/include/
+QMAKE_LIBDIR_QTOPIA	= $(QPEDIR)/lib
+
+QMAKE_LINK		= bfin-uclinux-g++ 
+QMAKE_LINK_SHLIB	= bfin-uclinux-g++
+QMAKE_LFLAGS		= -Wl,-elf2flt -mid-shared-library -mshared-library-id=0 -Wl,-shared-lib-id,0
+QMAKE_LFLAGS_RELEASE	= # -mid-shared-library -mshared-library-id=0
+QMAKE_LFLAGS_DEBUG	=
+QMAKE_LFLAGS_SHLIB      = -shared 
+QMAKE_LFLAGS_PLUGIN     = $$QMAKE_LFLAGS_SHLIB
+QMAKE_LFLAGS_SONAME     = -Wl,-soname,
+QMAKE_LFLAGS_THREAD     =
+QMAKE_RPATH             = -Wl,-rpath,
+
+QMAKE_LIBS		= 
+QMAKE_LIBS_DYNLOAD      = /usr/lib/libdl.a
+QMAKE_LIBS_X11		= 
+QMAKE_LIBS_X11SM	= 
+QMAKE_LIBS_QT		= -lqte
+QMAKE_LIBS_QT_THREAD    = -lqte-mt
+QMAKE_LIBS_QT_OPENGL	= -lqgl
+QMAKE_LIBS_QTOPIA	= -lqtopia -lqpe
+QMAKE_LIBS_THREAD       = -lpthread
+
+QMAKE_MOC		= $$[QT_INSTALL_BINS]/moc
+QMAKE_UIC		= $$[QT_INSTALL_BINS]/uic
+
+QMAKE_AR		= bfin-uclinux-ar cqs
+QMAKE_RANLIB		=
+
+QMAKE_TAR		= tar -cf
+QMAKE_GZIP		= gzip -9f
+
+QMAKE_COPY		= cp -f
+QMAKE_MOVE		= mv -f
+QMAKE_DEL_FILE		= rm -f
+QMAKE_DEL_DIR		= rmdir
+QMAKE_STRIP             = strip
+QMAKE_CHK_DIR_EXISTS	= test -d
+QMAKE_MKDIR		= mkdir -p
+include(../../common/unix.conf)
+load(qt_config)
+#/home/f/blackfin-toolchain-06r2/opt/uClinux/bfin-uclinux/bfin-uclinux/lib/libdl_pic.a
+#/home/f/blackfin-toolchain-06r2/opt/uClinux/bfin-uclinux/bfin-uclinux/lib/mfdpic/libdl.a
+#/home/f/blackfin-toolchain-06r2/opt/uClinux/bfin-uclinux/bfin-uclinux/lib/mfdpic/libdl.so
+#/home/f/blackfin-toolchain-06r2/opt/uClinux/bfin-uclinux/bfin-uclinux/runtime/lib/libdl-0.9.27.so
+#/home/f/blackfin-toolchain-06r2/opt/uClinux/bfin-uclinux/bfin-uclinux/runtime/lib/libdl.so.0
diff -Nur -x 'Makefile*' -x.libs -x'.*' -x'*.bak' -x '*.prl' -x '*.status' -x '*.pri' qtopia-core-opensource-src-4.2.2/mkspecs/qws/linux-bfin-g++/qmake.static.conf qtopia-core-opensource-src-4.2.2-newest-toolchain/mkspecs/qws/linux-bfin-g++/qmake.static.conf
--- qtopia-core-opensource-src-4.2.2/mkspecs/qws/linux-bfin-g++/qmake.static.conf	1970-01-01 01:00:00.000000000 +0100
+++ qtopia-core-opensource-src-4.2.2-newest-toolchain/mkspecs/qws/linux-bfin-g++/qmake.static.conf	2007-01-03 16:07:32.000000000 +0100
@@ -0,0 +1,91 @@
+#
+# qmake configuration for linux-g++
+#
+
+MAKEFILE_GENERATOR	= UNIX
+TEMPLATE		= app
+CONFIG			+= qt link_prl
+QT                      += core gui network
+QMAKE_INCREMENTAL_STYLE = sublib
+
+QMAKE_CC		= bfin-uclinux-gcc
+QMAKE_LEX		= flex
+QMAKE_LEXFLAGS		=
+QMAKE_YACC		= yacc
+QMAKE_YACCFLAGS		= -d
+QMAKE_CFLAGS		= -mfdpic -pipe -D__linux__ -DQT_NO_LIBRARY -DQT_NOMMU -DQ_NO_SEMAPHORE -DQT_QWS_NO_SHM -DQT_NO_QWS_MULTIPROCESS
+QMAKE_CFLAGS_WARN_ON	= -Wall -W
+QMAKE_CFLAGS_WARN_OFF	=
+QMAKE_CFLAGS_RELEASE	= -O2
+QMAKE_CFLAGS_DEBUG	= -g
+QMAKE_CFLAGS_SHLIB	= 
+QMAKE_CFLAGS_YACC	= -Wno-unused -Wno-parentheses
+QMAKE_CFLAGS_THREAD	= -D_REENTRANT
+QMAKE_CFLAGS_HIDESYMS   = -fvisibility=hidden
+
+QMAKE_CXX		= bfin-uclinux-g++
+QMAKE_CXXFLAGS		= $$QMAKE_CFLAGS -DQWS -fno-exceptions -fno-rtti
+QMAKE_CXXFLAGS_WARN_ON	= $$QMAKE_CFLAGS_WARN_ON
+QMAKE_CXXFLAGS_WARN_OFF	= $$QMAKE_CFLAGS_WARN_OFF
+QMAKE_CXXFLAGS_RELEASE	= $$QMAKE_CFLAGS_RELEASE
+QMAKE_CXXFLAGS_DEBUG	= $$QMAKE_CFLAGS_DEBUG
+QMAKE_CXXFLAGS_SHLIB	= $$QMAKE_CFLAGS_SHLIB
+QMAKE_CXXFLAGS_YACC	= $$QMAKE_CFLAGS_YACC
+QMAKE_CXXFLAGS_THREAD	= $$QMAKE_CFLAGS_THREAD
+QMAKE_CXXFLAGS_HIDESYMS = $$QMAKE_CFLAGS_HIDESYMS -fvisibility-inlines-hidden
+
+QMAKE_INCDIR		=
+QMAKE_LIBDIR		=
+QMAKE_INCDIR_X11	= 
+QMAKE_LIBDIR_X11	= 
+QMAKE_INCDIR_QT		= $$[QT_INSTALL_HEADERS]
+QMAKE_LIBDIR_QT		= $$[QT_INSTALL_LIBS]
+QMAKE_INCDIR_OPENGL	= /usr/X11R6/include
+QMAKE_LIBDIR_OPENGL	= /usr/X11R6/lib
+QMAKE_INCDIR_QTOPIA	= $(QPEDIR)/include/
+QMAKE_LIBDIR_QTOPIA	= $(QPEDIR)/lib
+
+QMAKE_LINK		= bfin-uclinux-g++ 
+QMAKE_LINK_SHLIB	= bfin-uclinux-g++
+QMAKE_LFLAGS		= -Wl,-elf2flt -mid-shared-library -mshared-library-id=0 -Wl,-shared-lib-id,0
+QMAKE_LFLAGS_RELEASE	= # -mid-shared-library -mshared-library-id=0
+QMAKE_LFLAGS_DEBUG	=
+QMAKE_LFLAGS_SHLIB      = -shared
+QMAKE_LFLAGS_PLUGIN     = $$QMAKE_LFLAGS_SHLIB
+QMAKE_LFLAGS_SONAME     = -Wl,-soname,
+QMAKE_LFLAGS_THREAD     =
+QMAKE_RPATH             = -Wl,-rpath,
+
+QMAKE_LIBS		= 
+QMAKE_LIBS_DYNLOAD      = 
+QMAKE_LIBS_X11		= 
+QMAKE_LIBS_X11SM	= 
+QMAKE_LIBS_QT		= -lqte
+QMAKE_LIBS_QT_THREAD    = -lqte-mt
+QMAKE_LIBS_QT_OPENGL	= -lqgl
+QMAKE_LIBS_QTOPIA	= -lqtopia -lqpe
+QMAKE_LIBS_THREAD       = -lpthread
+
+QMAKE_MOC		= $$[QT_INSTALL_BINS]/moc
+QMAKE_UIC		= $$[QT_INSTALL_BINS]/uic
+
+QMAKE_AR		= bfin-uclinux-ar cqs
+QMAKE_RANLIB		=
+
+QMAKE_TAR		= tar -cf
+QMAKE_GZIP		= gzip -9f
+
+QMAKE_COPY		= cp -f
+QMAKE_MOVE		= mv -f
+QMAKE_DEL_FILE		= rm -f
+QMAKE_DEL_DIR		= rmdir
+QMAKE_STRIP             = strip
+QMAKE_CHK_DIR_EXISTS	= test -d
+QMAKE_MKDIR		= mkdir -p
+include(../../common/unix.conf)
+load(qt_config)
+#/home/f/blackfin-toolchain-06r2/opt/uClinux/bfin-uclinux/bfin-uclinux/lib/libdl_pic.a
+#/home/f/blackfin-toolchain-06r2/opt/uClinux/bfin-uclinux/bfin-uclinux/lib/mfdpic/libdl.a
+#/home/f/blackfin-toolchain-06r2/opt/uClinux/bfin-uclinux/bfin-uclinux/lib/mfdpic/libdl.so
+#/home/f/blackfin-toolchain-06r2/opt/uClinux/bfin-uclinux/bfin-uclinux/runtime/lib/libdl-0.9.27.so
+#/home/f/blackfin-toolchain-06r2/opt/uClinux/bfin-uclinux/bfin-uclinux/runtime/lib/libdl.so.0
diff -Nur -x 'Makefile*' -x.libs -x'.*' -x'*.bak' -x '*.prl' -x '*.status' -x '*.pri' qtopia-core-opensource-src-4.2.2/mkspecs/qws/linux-bfin-g++/qplatformdefs.h qtopia-core-opensource-src-4.2.2-newest-toolchain/mkspecs/qws/linux-bfin-g++/qplatformdefs.h
--- qtopia-core-opensource-src-4.2.2/mkspecs/qws/linux-bfin-g++/qplatformdefs.h	1970-01-01 01:00:00.000000000 +0100
+++ qtopia-core-opensource-src-4.2.2-newest-toolchain/mkspecs/qws/linux-bfin-g++/qplatformdefs.h	2007-01-03 16:07:32.000000000 +0100
@@ -0,0 +1,145 @@
+/****************************************************************************
+**
+** Copyright (C) 1992-2006 Trolltech ASA. All rights reserved.
+**
+** This file is part of the qmake spec of the Qt Toolkit.
+**
+** This file may be used under the terms of the GNU General Public
+** License version 2.0 as published by the Free Software Foundation
+** and appearing in the file LICENSE.GPL included in the packaging of
+** this file.  Please review the following information to ensure GNU
+** General Public Licensing requirements will be met:
+** http://www.trolltech.com/products/qt/opensource.html
+**
+** If you are unsure which license is appropriate for your use, please
+** review the following information:
+** http://www.trolltech.com/products/qt/licensing.html or contact the
+** sales department at sales@trolltech.com.
+**
+** This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
+** WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
+**
+****************************************************************************/
+
+#ifndef QPLATFORMDEFS_H
+#define QPLATFORMDEFS_H
+
+// Get Qt defines/settings
+
+#include "qglobal.h"
+
+// Set any POSIX/XOPEN defines at the top of this file to turn on specific APIs
+
+// 1) need to reset default environment if _BSD_SOURCE is defined
+// 2) need to specify POSIX thread interfaces explicitly in glibc 2.0
+// 3) it seems older glibc need this to include the X/Open stuff
+#ifndef _GNU_SOURCE
+#  define _GNU_SOURCE
+#endif
+
+#include <unistd.h>
+
+
+// We are hot - unistd.h should have turned on the specific APIs we requested
+
+#include <features.h>
+#include <pthread.h>
+#include <dirent.h>
+#include <fcntl.h>
+#include <grp.h>
+#include <pwd.h>
+#include <signal.h>
+#include <dlfcn.h>
+
+#include <sys/types.h>
+#include <sys/ioctl.h>
+#include <sys/ipc.h>
+#include <sys/time.h>
+#include <sys/shm.h>
+#include <sys/socket.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <netinet/in.h>
+#ifndef QT_NO_IPV6IFNAME
+#include <net/if.h>
+#endif
+
+#ifdef QT_LARGEFILE_SUPPORT
+#define QT_STATBUF              struct stat64
+#define QT_STATBUF4TSTAT        struct stat64
+#define QT_STAT                 ::stat64
+#define QT_FSTAT                ::fstat64
+#define QT_LSTAT                ::lstat64
+#define QT_OPEN                 ::open64
+#define QT_TRUNCATE             ::truncate64
+#define QT_FTRUNCATE            ::ftruncate64
+#define QT_LSEEK                ::lseek64
+#else
+#define QT_STATBUF              struct stat
+#define QT_STATBUF4TSTAT        struct stat
+#define QT_STAT                 ::stat
+#define QT_FSTAT                ::fstat
+#define QT_LSTAT                ::lstat
+#define QT_OPEN                 ::open
+#define QT_TRUNCATE             ::truncate
+#define QT_FTRUNCATE            ::ftruncate
+#define QT_LSEEK                ::lseek
+#endif
+
+#ifdef QT_LARGEFILE_SUPPORT
+#define QT_FOPEN                ::fopen64
+#define QT_FSEEK                ::fseeko64
+#define QT_FTELL                ::ftello64
+#define QT_FGETPOS              ::fgetpos64
+#define QT_FSETPOS              ::fsetpos64
+#define QT_FPOS_T               fpos64_t
+#define QT_OFF_T                off64_t
+#else
+#define QT_FOPEN                ::fopen
+#define QT_FSEEK                ::fseek
+#define QT_FTELL                ::ftell
+#define QT_FGETPOS              ::fgetpos
+#define QT_FSETPOS              ::fsetpos
+#define QT_FPOS_T               fpos_t
+#define QT_OFF_T                long
+#endif
+
+#define QT_STAT_REG		S_IFREG
+#define QT_STAT_DIR		S_IFDIR
+#define QT_STAT_MASK		S_IFMT
+#define QT_STAT_LNK		S_IFLNK
+#define QT_SOCKET_CONNECT	::connect
+#define QT_SOCKET_BIND		::bind
+#define QT_FILENO		fileno
+#define QT_CLOSE		::close
+#define QT_READ			::read
+#define QT_WRITE		::write
+#define QT_ACCESS		::access
+#define QT_GETCWD		::getcwd
+#define QT_CHDIR		::chdir
+#define QT_MKDIR		::mkdir
+#define QT_RMDIR		::rmdir
+#define QT_OPEN_RDONLY		O_RDONLY
+#define QT_OPEN_WRONLY		O_WRONLY
+#define QT_OPEN_RDWR		O_RDWR
+#define QT_OPEN_CREAT		O_CREAT
+#define QT_OPEN_TRUNC		O_TRUNC
+#define QT_OPEN_APPEND		O_APPEND
+
+#define QT_SIGNAL_RETTYPE	void
+#define QT_SIGNAL_ARGS		int
+#define QT_SIGNAL_IGNORE	SIG_IGN
+
+#if defined(__GLIBC__) && (__GLIBC__ >= 2)
+#define QT_SOCKLEN_T		socklen_t
+#else
+#define QT_SOCKLEN_T		int
+#endif
+
+#if defined(_XOPEN_SOURCE) && (_XOPEN_SOURCE >= 500)
+#define QT_SNPRINTF		::snprintf
+#define QT_VSNPRINTF		::vsnprintf
+#endif
+
+
+#endif // QPLATFORMDEFS_H
diff -Nur -x 'Makefile*' -x.libs -x'.*' -x'*.bak' -x '*.prl' -x '*.status' -x '*.pri' qtopia-core-opensource-src-4.2.2/src/3rdparty/zlib/configure qtopia-core-opensource-src-4.2.2-newest-toolchain/src/3rdparty/zlib/configure
--- qtopia-core-opensource-src-4.2.2/src/3rdparty/zlib/configure	2006-11-27 18:47:30.000000000 +0100
+++ qtopia-core-opensource-src-4.2.2-newest-toolchain/src/3rdparty/zlib/configure	2007-01-03 16:07:35.000000000 +0100
@@ -414,7 +414,7 @@
 #include <sys/mman.h>
 #include <sys/stat.h>
 caddr_t hello() {
-  return mmap((caddr_t)0, (off_t)0, PROT_READ, MAP_SHARED, 0, (off_t)0);
+  return mmap((caddr_t)0, (off_t)0, PROT_READ, MAP_PRIVATE, 0, (off_t)0);
 }
 EOF
 if test "`($CC -c $CFLAGS $test.c) 2>&1`" = ""; then
diff -Nur -x 'Makefile*' -x.libs -x'.*' -x'*.bak' -x '*.prl' -x '*.status' -x '*.pri' qtopia-core-opensource-src-4.2.2/src/3rdparty/zlib/minigzip.c qtopia-core-opensource-src-4.2.2-newest-toolchain/src/3rdparty/zlib/minigzip.c
--- qtopia-core-opensource-src-4.2.2/src/3rdparty/zlib/minigzip.c	2006-11-27 18:47:30.000000000 +0100
+++ qtopia-core-opensource-src-4.2.2-newest-toolchain/src/3rdparty/zlib/minigzip.c	2007-01-03 16:07:35.000000000 +0100
@@ -145,7 +145,7 @@
     if (buf_len <= 0) return Z_ERRNO;
 
     /* Now do the actual mmap: */
-    buf = mmap((caddr_t) 0, buf_len, PROT_READ, MAP_SHARED, ifd, (off_t)0);
+    buf = mmap((caddr_t) 0, buf_len, PROT_READ, MAP_PRIVATE, ifd, (off_t)0);
     if (buf == (caddr_t)(-1)) return Z_ERRNO;
 
     /* Compress the whole file at once: */
diff -Nur -x 'Makefile*' -x.libs -x'.*' -x'*.bak' -x '*.prl' -x '*.status' -x '*.pri' qtopia-core-opensource-src-4.2.2/src/corelib/global/qconfig.h.qmake qtopia-core-opensource-src-4.2.2-newest-toolchain/src/corelib/global/qconfig.h.qmake
--- qtopia-core-opensource-src-4.2.2/src/corelib/global/qconfig.h.qmake	1970-01-01 01:00:00.000000000 +0100
+++ qtopia-core-opensource-src-4.2.2-newest-toolchain/src/corelib/global/qconfig.h.qmake	2007-02-16 16:22:43.000000000 +0100
@@ -0,0 +1 @@
+/* All features enabled while building qmake */
diff -Nur -x 'Makefile*' -x.libs -x'.*' -x'*.bak' -x '*.prl' -x '*.status' -x '*.pri' qtopia-core-opensource-src-4.2.2/src/corelib/io/qfilesystemwatcher_inotify.cpp qtopia-core-opensource-src-4.2.2-newest-toolchain/src/corelib/io/qfilesystemwatcher_inotify.cpp
--- qtopia-core-opensource-src-4.2.2/src/corelib/io/qfilesystemwatcher_inotify.cpp	2006-11-27 18:47:33.000000000 +0100
+++ qtopia-core-opensource-src-4.2.2-newest-toolchain/src/corelib/io/qfilesystemwatcher_inotify.cpp	2007-01-03 16:07:35.000000000 +0100
@@ -89,7 +89,10 @@
 # define __NR_inotify_add_watch 270
 # define __NR_inotify_rm_watch  271
 #else
-# error "This architecture is not supported. Please talk to qt-bugs@trolltech.com"
+//# error "This architecture is not supported. Please talk to qt-bugs@trolltech.com"
+# define __NR_inotify_init      316
+# define __NR_inotify_add_watch 317
+# define __NR_inotify_rm_watch  318
 #endif
 
 #ifdef QT_LSB
diff -Nur -x 'Makefile*' -x.libs -x'.*' -x'*.bak' -x '*.prl' -x '*.status' -x '*.pri' qtopia-core-opensource-src-4.2.2/src/corelib/io/qprocess_unix.cpp qtopia-core-opensource-src-4.2.2-newest-toolchain/src/corelib/io/qprocess_unix.cpp
--- qtopia-core-opensource-src-4.2.2/src/corelib/io/qprocess_unix.cpp	2006-11-27 18:47:33.000000000 +0100
+++ qtopia-core-opensource-src-4.2.2-newest-toolchain/src/corelib/io/qprocess_unix.cpp	2007-01-03 16:07:35.000000000 +0100
@@ -541,7 +541,11 @@
 
     QByteArray encodedProg = QFile::encodeName(program);
     processManager()->lock();
+#ifdef QT_NOMMU
+    pid_t childPid = vfork();
+#else
     pid_t childPid = fork();
+#endif
     if (childPid < 0) {
         // Cleanup, report error and return
         processManager()->unlock();
@@ -1149,14 +1153,22 @@
     int startedPipe[2];
     ::pipe(startedPipe);
 
+#ifdef QT_NOMMU
+    pid_t childPid = vfork();
+#else
     pid_t childPid = fork();
+#endif
     if (childPid == 0) {
         ::setsid();
         ::signal(SIGHUP, SIG_IGN);
         qt_native_close(startedPipe[0]);
         ::signal(SIGPIPE, SIG_DFL);
 
+#ifdef QT_NOMMU
+        pid_t doubleForkPid = vfork();
+#else
         pid_t doubleForkPid = fork();
+#endif
         if (doubleForkPid == 0) {
             ::fcntl(startedPipe[1], F_SETFD, FD_CLOEXEC);
 
diff -Nur -x 'Makefile*' -x.libs -x'.*' -x'*.bak' -x '*.prl' -x '*.status' -x '*.pri' qtopia-core-opensource-src-4.2.2/src/corelib/tools/qlocale.cpp qtopia-core-opensource-src-4.2.2-newest-toolchain/src/corelib/tools/qlocale.cpp
--- qtopia-core-opensource-src-4.2.2/src/corelib/tools/qlocale.cpp	2006-11-27 18:47:34.000000000 +0100
+++ qtopia-core-opensource-src-4.2.2-newest-toolchain/src/corelib/tools/qlocale.cpp	2007-01-03 16:07:35.000000000 +0100
@@ -48,8 +48,10 @@
 #include <qdebug.h>
 
 #if defined(Q_OS_LINUX) && !defined(__UCLIBC__)
+#ifndef QT_NOMMU
 #    include <fenv.h>
 #endif
+#endif
 
 #if !defined(QT_QLOCALE_NEEDS_VOLATILE)
 #  if defined(Q_CC_GNU)
@@ -5651,9 +5653,11 @@
 #endif
 
 #if defined(Q_OS_LINUX) && !defined(__UCLIBC__)
+#ifndef QT_NOMMU
     fenv_t envp;
     feholdexcept(&envp);
 #endif
+#endif
 
     char *s = _qdtoa(d, mode, ndigits, decpt, sign, rve, resultp);
 
@@ -5667,8 +5671,10 @@
 #endif //Q_OS_WIN
 
 #if defined(Q_OS_LINUX) && !defined(__UCLIBC__)
+#ifndef QT_NOMMU
     fesetenv(&envp);
 #endif
+#endif
 
     return s;
 }
diff -Nur -x 'Makefile*' -x.libs -x'.*' -x'*.bak' -x '*.prl' -x '*.status' -x '*.pri' qtopia-core-opensource-src-4.2.2/src/gui/embedded/qlock.cpp qtopia-core-opensource-src-4.2.2-newest-toolchain/src/gui/embedded/qlock.cpp
--- qtopia-core-opensource-src-4.2.2/src/gui/embedded/qlock.cpp	2006-11-27 18:47:38.000000000 +0100
+++ qtopia-core-opensource-src-4.2.2-newest-toolchain/src/gui/embedded/qlock.cpp	2007-01-03 16:07:39.000000000 +0100
@@ -28,10 +28,12 @@
 #include "qwssignalhandler_p.h"
 #include <unistd.h>
 #include <sys/types.h>
+#include <fcntl.h>
 #if defined(Q_OS_DARWIN)
 #define Q_NO_SEMAPHORE
 #include <sys/stat.h>
 #include <sys/file.h>
+#include <fcntl.h>
 #else
 #include <sys/sem.h>
 #if defined(__GNU_LIBRARY__) && !defined(_SEM_SEMUN_UNDEFINED) \
@@ -47,6 +49,7 @@
 #endif
 #endif
 #include <sys/ipc.h>
+#include <sys/file.h>
 #include <string.h>
 #include <errno.h>
 #include <qdebug.h>
diff -Nur -x 'Makefile*' -x.libs -x'.*' -x'*.bak' -x '*.prl' -x '*.status' -x '*.pri' qtopia-core-opensource-src-4.2.2/src/gui/embedded/qscreenlinuxfb_qws.cpp qtopia-core-opensource-src-4.2.2-newest-toolchain/src/gui/embedded/qscreenlinuxfb_qws.cpp
--- qtopia-core-opensource-src-4.2.2/src/gui/embedded/qscreenlinuxfb_qws.cpp	2006-11-27 18:47:38.000000000 +0100
+++ qtopia-core-opensource-src-4.2.2-newest-toolchain/src/gui/embedded/qscreenlinuxfb_qws.cpp	2007-01-03 16:07:39.000000000 +0100
@@ -355,7 +355,7 @@
     mapsize=finfo.smem_len;
 
     data = (unsigned char *)mmap(0, mapsize, PROT_READ | PROT_WRITE,
-                                 MAP_SHARED, d_ptr->fd, 0);
+                                 MAP_PRIVATE, d_ptr->fd, 0);
 
     if ((long)data == -1) {
         perror("QLinuxFbScreen::connect");
diff -Nur -x 'Makefile*' -x.libs -x'.*' -x'*.bak' -x '*.prl' -x '*.status' -x '*.pri' qtopia-core-opensource-src-4.2.2/src/gui/painting/qpdf.cpp qtopia-core-opensource-src-4.2.2-newest-toolchain/src/gui/painting/qpdf.cpp
--- qtopia-core-opensource-src-4.2.2/src/gui/painting/qpdf.cpp	2006-11-27 18:47:35.000000000 +0100
+++ qtopia-core-opensource-src-4.2.2-newest-toolchain/src/gui/painting/qpdf.cpp	2007-01-03 16:07:39.000000000 +0100
@@ -1396,11 +1396,16 @@
             return false;
         }
 
+#ifdef QT_NOMMU
+        pid_t pid = vfork();
+#else
         pid_t pid = fork();
+#endif
         if (pid == 0) {       // child process
             // if possible, exit quickly, so the actual lp/lpr
             // becomes a child of init, and ::waitpid() is
             // guaranteed not to wait.
+#ifndef QT_NOMMU
             if (fork() > 0) {
                 closeAllOpenFds();
 
@@ -1412,6 +1417,7 @@
                 (void)execl("/usr/bin/true", "true", (char *)0);
                 ::exit(0);
             }
+#endif
             dup2(fds[0], 0);
 
             closeAllOpenFds();
diff -Nur -x 'Makefile*' -x.libs -x'.*' -x'*.bak' -x '*.prl' -x '*.status' -x '*.pri' qtopia-core-opensource-src-4.2.2/src/qt3support/other/q3process_unix.cpp qtopia-core-opensource-src-4.2.2-newest-toolchain/src/qt3support/other/q3process_unix.cpp
--- qtopia-core-opensource-src-4.2.2/src/qt3support/other/q3process_unix.cpp	2006-11-27 18:47:42.000000000 +0100
+++ qtopia-core-opensource-src-4.2.2-newest-toolchain/src/qt3support/other/q3process_unix.cpp	2007-01-03 16:07:48.000000000 +0100
@@ -777,7 +777,11 @@
 
     // fork and exec
     QApplication::flushX();
+#ifdef QT_NOMMU
+    pid_t pid = vfork();
+#else
     pid_t pid = fork();
+#endif
     if ( pid == 0 ) {
 	// child
 	d->closeOpenSocketsForChild();
