--- mediastreamer2/configure.orig	2007-10-16 10:58:55.000000000 +0800
+++ mediastreamer2/configure	2007-10-15 18:24:12.000000000 +0800
@@ -20058,9 +20058,9 @@
 	CFLAGS="$CFLAGS -D_BIGENDIAN "
 fi
 
-if test $GCC = yes && test $wall_werror = yes;  then
-	CFLAGS="$CFLAGS -Werror "
-fi
+#if test $GCC = yes && test $wall_werror = yes;  then
+#	CFLAGS="$CFLAGS -Werror "
+#fi
 
 case $OS in
   darwin*)
