turn wordsize check from a runtime test to compile time

--- MPlayer-1.0rc1/configure
+++ MPlayer-1.0rc1/configure
@@ -2901,12 +2912,29 @@
 
 echocheck "word size"
 _mp_wordsize="#undef MP_WORDSIZE"
-cat > $TMPC << EOF
-#include <stdio.h>
+_wordsize=0
+while true; do
+  cat > $TMPC << EOF
+#include <stdlib.h>
 #include <sys/types.h>
-int main(void) { printf("%d\n", sizeof(size_t)*8); return 0; }
+int main(void) {
+  static int test_array [1 - 2 * !(((long int) (sizeof (size_t))) <= $_wordsize)];
+  test_array [0] = 0;
+  return 0;
+}
 EOF
-cc_check && _wordsize=`$TMPO` && _mp_wordsize="#define MP_WORDSIZE $_wordsize"
+  if cc_check; then
+    break
+  fi
+  # sanity check
+  if test $_wordsize -gt 20; then
+    _wordsize=""
+    break
+  fi
+  _wordsize=`expr $_wordsize + 1`
+done
+_wordsize=`expr $_wordsize '*' 8`
+[ -n "$_wordsize" ] && _mp_wordsize="#define MP_WORDSIZE `expr $_wordsize '*' 8`"
-echores "$_wordsize"
+echores "$_wordsize bits" 
 
 
